File: core/src/main/java/com/illuzionzstudios/customfishing/reward/loader/FishingRewardLoader.java
package com.illuzionzstudios.customfishing.reward.loader;

import com.cryptomorin.xseries.XSound;
import com.illuzionzstudios.customfishing.reward.FishingReward;
import com.illuzionzstudios.mist.config.ConfigSection;
import com.illuzionzstudios.mist.config.locale.MistString;
import com.illuzionzstudios.mist.config.serialization.loader.YamlFileLoader;
import com.illuzionzstudios.mist.exception.PluginException;

import java.util.ArrayList;

/**
 * Disk file loader for a {@link com.illuzionzstudios.customfishing.reward.FishingReward}. Takes YAML files
 */
public class FishingRewardLoader extends YamlFileLoader<FishingReward> {

    public FishingRewardLoader(String directory, String fileName) {
        super(directory, fileName);
    }

    @Override
    public void saveYaml() {
        // TODO: Implement saving
    }

    @Override
    public FishingReward loadYamlObject() {
        // Lets try build the reward
        FishingReward.FishingRewardBuilder builder = FishingReward.builder();

        // Cause to send if fails
        String cause = "Failed to load reward file";

        // If fails throws error with cause message
        // set before trying to set builder
        try {
            // Load everything
            cause = "Could not load name";
            builder.name(config.getString("name"));

            cause = "Could not load commands";
            builder.commands(config.getStringList("commands"));

            cause = "Could not load items";
            ArrayList<FishingItemLoader> items = new ArrayList<>();
            // Parse custom items
            for (ConfigSection section : config.getSections("items")) {
                items.add(new FishingItemLoader(section));
            }
            builder.items(items);

            cause = "Could not load messages";
            ArrayList<MistString> messages = new ArrayList<>();
            config.getStringList("messages").forEach(string -> messages.add(new MistString(string)));
            builder.messages(messages);

            cause = "Could not load broadcasts";
            ArrayList<MistString> broadcasts = new ArrayList<>();
            config.getStringList("broadcasts").forEach(string -> broadcasts.add(new MistString(string)));
            builder.broadcasts(broadcasts);

            cause = "Could not load title";
            builder.title(new MistString(config.getString("title")));

            cause = "Could not load sub title";
            builder.subtitle(new MistString(config.getString("sub-title")));

            cause = "Could not load chance";
            builder.chance(config.getDouble("weight"));

            cause = "Could not load if to give vanilla rewards";
            builder.vanillaRewards(config.getBoolean("vanilla-rewards"));

            cause = "Could not load exp to give";
            builder.experienceRange(config.get("exp-amount") != null ? config.get("exp-amount").toString() : "0");

            // Set sound
            cause = "Sound " + config.getString("sound") + " is not valid";
            try {
                builder.sound(XSound.matchXSound(config.getString("sound")).get());
            } catch (Exception e) {
                builder.sound(null);
            }

            cause = "Could not load permission";
            builder.permission(config.getString("requirements.permission"));

            cause = "Could not load worlds";
            builder.worlds(config.getStringList("requirements.worlds"));

            cause = "Could not load regions";
            builder.regions(config.getStringList("requirements.regions"));

            cause = "Could not load blocked regions";
            builder.blockedRegions(config.getStringList("requirements.blocked-regions"));
        } catch (final Throwable ex) {
            throw new PluginException(cause);
        }

        return builder.build();
    }
}

File: src/main/kotlin/com/illuzionzstudios/customfishing/reward/fishing/loader/FishingRewardLoader.kt
package com.illuzionzstudios.customfishing.reward.fishing.loader

import com.cryptomorin.xseries.XSound
import com.illuzionzstudios.customfishing.reward.fishing.FishingReward
import com.illuzionzstudios.customfishing.reward.fishing.item.FishingItem
import com.illuzionzstudios.mist.config.ConfigSection
import com.illuzionzstudios.mist.config.YamlConfig
import com.illuzionzstudios.mist.config.locale.MistString
import com.illuzionzstudios.mist.config.serialization.loader.YamlFileLoader
import com.illuzionzstudios.mist.requirement.PlayerRequirement
import com.illuzionzstudios.mist.requirement.PlayerRequirementLoader

/**
 * Load a fishing reward from a file
 */
class FishingRewardLoader(directory: String, fileName: String) : YamlFileLoader<FishingReward>(directory, fileName) {

    override fun loadYamlObject(file: YamlConfig?): FishingReward {
        val reward = FishingReward()

        reward.name = file?.getString("name")
        reward.commands = file?.getStringList("commands")

        val items: MutableList<FishingItem> = ArrayList()
        file?.getSections("items")?.forEach {
            items.add(FishingItemLoader(it!!).`object`)
        }
        reward.items = items

        reward.messages = file?.getStringList("messages")
        reward.broadcasts = file?.getStringList("broadcasts")

        reward.title = file?.getString("title")
        reward.subtitle = file?.getString("sub-title")

        reward.chance = file?.getDouble("weight") ?: 0.0
        reward.vanillaRewards = file?.getBoolean("vanilla-rewards") ?: false
        reward.experienceRange = file?.getString("exp-amount") ?: "1 to 6"
        try {
            reward.sound = XSound.matchXSound(config?.getString("sound") ?: "").get()
        } catch (ignored: Exception) {
        }

        // Load requirements
        val requirements: MutableList<PlayerRequirement> = ArrayList()
        file?.getSections("requirements")?.forEach {
            requirements.add(PlayerRequirementLoader(it!!).`object`)
        }
        reward.requirements = requirements

        return reward
    }

    override fun saveYaml() {
        TODO("Not yet implemented")
    }
}